src/Protocols/Df.hs

```diff
@@ -413,54 +536,217 @@ roundrobin ::
   Circuit (Df dom a) (Df dom a)
-registerBwd = DfLike.registerBwd Proxy
+registerBwd
+  = forceResetSanity |> Circuit (C.mealyB go (NoData, NoData))
+ where
+   go (ra0, rb) (iDat, Ack iAck) =
+     (s, (Ack oAck, rb))
+    where
+     oAck = Maybe.isNothing (dataToMaybe ra0)
+     ra1 = if oAck then iDat else ra0
+     s = if Maybe.isNothing (dataToMaybe rb) || iAck then (NoData, ra1) else (ra1, rb)
+
+-- | A fifo buffer with user-provided depth. Uses blockram to store data. Can
+-- handle simultaneous write and read (full throughput rate).
+fifo ::
+  forall dom a depth.
+  (C.HiddenClockResetEnable dom, C.KnownNat depth, C.NFDataX a, 1 C.<= depth) =>
+  C.SNat depth ->
+  Circuit (Df dom a) (Df dom a)
+fifo fifoDepth = Circuit $ C.hideReset circuitFunction where
+
+  -- implemented using a fixed-size array
+  --   write location and read location are both stored
+  --   to write, write to current location and move one to the right
+  --   to read, read from current location and move one to the right
+  --   loop around from the end to the beginning if necessary
+
+  circuitFunction reset (inpA, inpB) = (otpA, otpB) where
+    -- initialize bram
+    brRead = C.readNew
+             (C.blockRamU C.NoClearOnReset fifoDepth (C.errorX "No reset function"))
+             brReadAddr brWrite
+    -- run the state machine (a mealy machine)
+    (brReadAddr, brWrite, otpA, otpB)
+      = C.unbundle
+      $ C.mealy machineAsFunction s0
+      $ C.bundle (brRead, C.unsafeToHighPolarity reset, inpA, inpB)
+
+  -- when reset is on, set state to initial state and output blank outputs
+  machineAsFunction _ (_, True, _, _) = (s0, (0, Nothing, Ack False, NoData))
+  machineAsFunction (rAddr,wAddr,amtLeft) (brRead, False, pushData, Ack popped) =
+    let -- potentially push an item onto blockram
+        maybePush = if amtLeft > 0 then dataToMaybe pushData else Nothing
+        brWrite = (wAddr,) <$> maybePush
+        -- adjust write address and amount left
+        --   (output state machine doesn't see amountLeft')
+        (wAddr', amtLeft') = if (Maybe.isJust maybePush)
+                             then (C.satSucc C.SatWrap wAddr, amtLeft-1)
+                             else (wAddr, amtLeft)
+        -- if we're about to push onto an empty queue, we can pop immediately instead
+        (brRead_, amtLeft_) = if (amtLeft == maxBound && Maybe.isJust maybePush)
+                              then (Maybe.fromJust maybePush, amtLeft')
+                              else (brRead, amtLeft)
+        -- adjust blockram read address and amount left
+        (rAddr', amtLeft'') = if (amtLeft_ < maxBound && popped)
+                              then (C.satSucc C.SatWrap rAddr, amtLeft'+1)
+                              else (rAddr, amtLeft')
+        brReadAddr = rAddr'
+        -- return our new state and outputs
+        otpAck = Maybe.isJust maybePush
+        otpDat = if amtLeft_ < maxBound then Data brRead_ else NoData
```

Comment:
```suggestion
        otpDat = if amtLeft_ < maxBound then Data brRead_ else NoData
```

Try to avoid the `variable_` pattern, as it's easy to miss when reading code.